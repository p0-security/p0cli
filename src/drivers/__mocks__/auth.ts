export const authenticate = async () => ({
  identity: {
    credential: {
      access_token: "test-access-token",
    },
    org: {
      ssoProvider: "oidc-pkce",
      providerDomain: "test.okta.com",
      providerType: "okta",
      slug: "test-org",
      tenantId: "test-tenant",
    },
  },
  userCredential: {
    user: {
      tenantId: "test-tenant",
    },
  },
});

export const cached = async (_label: string, callback: () => Promise<any>) =>
  await callback();
