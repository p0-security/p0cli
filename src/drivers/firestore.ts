/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { Identity } from "../types/identity";
import { getContactMessage, loadConfig } from "./config";
import { bootstrapConfig } from "./env";
import { print2 } from "./stdio";
import { EXPIRED_CREDENTIALS_MESSAGE } from "./util";
import { FirebaseApp, FirebaseError, initializeApp } from "firebase/app";
import {
  getAuth,
  OAuthProvider,
  SignInMethod,
  signInWithCredential,
  UserCredential,
} from "firebase/auth";
import {
  collection as fsCollection,
  CollectionReference,
  doc as fsDoc,
  DocumentReference,
  getFirestore,
  terminate,
  Firestore,
} from "firebase/firestore";

const bootstrapApp = initializeApp(bootstrapConfig.fs, "bootstrapApp");
const bootstrapFirestore = getFirestore(bootstrapApp);

let app: FirebaseApp;
let firestore: Firestore;

export async function initializeFirebase() {
  if (!firestore) {
    const tenantConfig = await loadConfig();
    app = initializeApp(tenantConfig.fs, "authFirebase");
    firestore = getFirestore(app);
  }
}

const findProviderId = (identity: Identity) => {
  switch (identity.org.ssoProvider) {
    case "google":
      return SignInMethod.GOOGLE;
    case "google-oidc":
      return "oidc.google-oidc";
    default:
      return identity.org.providerId;
  }
};

export async function authenticateToFirebase(
  identity: Identity,
  options?: {
    debug?: boolean;
  }
): Promise<UserCredential> {
  const { credential } = identity;
  const tenantId = identity.org.tenantId;

  await initializeFirebase();

  const provider = new OAuthProvider(findProviderId(identity));

  const firebaseCredential = provider.credential({
    accessToken: credential.access_token,
    idToken: credential.id_token,
  });

  const auth = getAuth(app);
  auth.tenantId = tenantId;

  let userCredential;
  try {
    userCredential = await signInWithCredential(auth, firebaseCredential);
  } catch (error) {
    if (
      error instanceof FirebaseError &&
      error.code === "auth/invalid-credential"
    ) {
      throw EXPIRED_CREDENTIALS_MESSAGE;
    } else {
      if (options?.debug) {
        print2("Authentication error: " + error);
      }
      throw `An unexpected error occurred during authentication.\n${getContactMessage()}`;
    }
  }

  if (!userCredential?.user?.email) {
    throw `Can not sign in: this user has previously signed in with a different identity provider.\n${getContactMessage()}`;
  }

  return userCredential;
}

export const collection = <T>(path: string, ...pathSegments: string[]) => {
  return fsCollection(
    firestore,
    path,
    ...pathSegments
  ) as CollectionReference<T>;
};

export const doc = <T>(path: string) => {
  return fsDoc(firestore, path) as DocumentReference<T>;
};

export const bootstrapDoc = <T>(path: string) => {
  return fsDoc(bootstrapFirestore, path) as DocumentReference<T>;
};

/** Ensures that Firestore is shutdown at command termination
 *
 * This prevents Firestore from holding the command on execution completion or failure.
 */
export const fsShutdownGuard =
  <P, T>(cb: (args: P) => Promise<T>) =>
  async (args: P) => {
    try {
      await cb(args);
    } finally {
      shutdownFirebase();
    }
  };

export const shutdownFirebase = () => {
  if (bootstrapFirestore) void terminate(bootstrapFirestore);
  if (firestore) void terminate(firestore);
};
