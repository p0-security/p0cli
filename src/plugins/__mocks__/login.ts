export const pluginLoginMap = {
  google: jest.fn().mockResolvedValue({
    access_token: "test-access-token",
    id_token: "test-id-token",
    token_type: "oidc",
    scope: "oidc",
    expires_in: 3600,
    refresh_token: "test-refresh-token",
    device_secret: "test-device-secret",
  }),
};
