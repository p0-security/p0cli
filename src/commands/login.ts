/** Copyright Â© 2024-present P0 Security 

This file is part of @p0security/p0cli

@p0security/p0cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/p0cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/p0cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { IDENTITY_FILE_PATH, authenticate } from "../drivers/auth";
import { doc, guard } from "../drivers/firestore";
import { print2 } from "../drivers/stdio";
import { pluginLoginMap } from "../plugins/login";
import { TokenResponse } from "../types/oidc";
import { OrgData } from "../types/org";
import { getDoc } from "firebase/firestore";
import * as fs from "fs/promises";
import * as path from "path";
import yargs from "yargs";

/** Logs in the user
 *
 * Currently only supports login to a single organization. Login credentials, together
 * with organization details, are saved to {@link IDENTITY_FILE_PATH}.
 */
export const login = async (
  args: { org: string },
  options?: { skipAuthenticate?: boolean }
) => {
  const orgDoc = await getDoc<Omit<OrgData, "slug">, object>(
    doc(`orgs/${args.org}`)
  );
  const orgData = orgDoc.data();
  if (!orgData) throw "Could not find organization";

  const orgWithSlug: OrgData = { ...orgData, slug: args.org };

  const plugin = orgWithSlug?.ssoProvider;
  const loginFn = pluginLoginMap[plugin];

  if (!loginFn) throw "Unsupported login for your organization";

  const tokenResponse = await loginFn(orgWithSlug);
  await writeIdentity(orgWithSlug, tokenResponse);

  // validate auth
  if (!options?.skipAuthenticate) await authenticate({ noRefresh: true });

  print2(`You are now logged in, and can use the p0 CLI.`);
};

const writeIdentity = async (org: OrgData, credential: TokenResponse) => {
  const expires_at = Date.now() * 1e-3 + credential.expires_in - 1; // Add 1 second safety margin
  print2(`Saving authorization to ${IDENTITY_FILE_PATH}.`);
  const dir = path.dirname(IDENTITY_FILE_PATH);
  await fs.mkdir(dir, { recursive: true });
  await fs.writeFile(
    IDENTITY_FILE_PATH,
    JSON.stringify(
      {
        credential: { ...credential, expires_at },
        org,
      },
      null,
      2
    ),
    {
      mode: "600",
    }
  );
};

export const loginCommand = (yargs: yargs.Argv) =>
  yargs.command<{ org: string }>(
    "login <org>",
    "Log in to p0 using a web browser",
    (yargs) =>
      yargs.positional("org", {
        demandOption: true,
        type: "string",
        describe: "Your P0 organization ID",
      }),
    guard(login)
  );
