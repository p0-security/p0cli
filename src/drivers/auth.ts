import { login } from "../commands/login";
import { Identity } from "../types/identity";
import { auth } from "./firestore";
import {
  OAuthProvider,
  SignInMethod,
  signInWithCredential,
} from "firebase/auth";
import * as fs from "fs/promises";
import * as os from "os";
import * as path from "path";

export const IDENTITY_FILE_PATH = path.join(
  os.homedir(),
  ".p0",
  "identity.json"
);

export const loadCredentials = async (options?: { noRefresh?: boolean }) => {
  try {
    const buffer = await fs.readFile(IDENTITY_FILE_PATH);
    const identity: Identity = JSON.parse(buffer.toString());
    if (
      !options?.noRefresh &&
      identity.credential.expires_at < Date.now() * 1e-3
    ) {
      await login({ org: identity.org.slug }, { skipAuthenticate: true });
      console.error("\u200B"); // Force a new line
    }
    return identity;
  } catch (error: any) {
    if (error?.code === "ENOENT") {
      throw "Please run `p0 login <organization>` to use the P0 CLI.";
    }
    throw error;
  }
};

export const authenticate = async (options?: { noRefresh?: boolean }) => {
  const identity = await loadCredentials(options);
  const { credential } = identity;
  // TODO: Move to map lookup
  const provider = new OAuthProvider(
    identity.org.ssoProvider === "google"
      ? SignInMethod.GOOGLE
      : identity.org.providerId
  );
  const firebaseCredential = provider.credential({
    accessToken: credential.access_token,
    idToken: credential.id_token,
  });
  auth.tenantId = identity.org.tenantId;
  const userCredential = await signInWithCredential(auth, firebaseCredential);
  return { userCredential, identity };
};
