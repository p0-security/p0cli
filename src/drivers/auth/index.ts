/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { login } from "../../commands/login";
import { Authn, Identity } from "../../types/identity";
import { TokenResponse } from "../../types/oidc";
import { OrgData } from "../../types/org";
import { authenticateToFirebase } from "../firestore";
import { print2 } from "../stdio";
import { getIdentityCachePath, getIdentityFilePath } from "./path";
import * as fs from "fs/promises";
import * as os from "os";
import * as path from "path";

export const cached = async <T>(
  name: string,
  loader: () => Promise<T>,
  options: { duration: number },
  hasExpired?: (data: T) => boolean
): Promise<T> => {
  const identityCachePath = getIdentityCachePath();

  // Following lines sanitize input
  // nosemgrep: javascript.lang.security.audit.path-traversal.path-join-resolve-traversal.path-join-resolve-traversal
  const loc = path.resolve(path.join(identityCachePath, `${name}.json`));
  if (!loc.startsWith(identityCachePath)) {
    throw new Error("Illegal path traversal");
  }

  const loadCache = async () => {
    const data = await loader();
    if (!data) throw `Could not load credentials for "${name}"`;
    await fs.mkdir(path.dirname(loc), { recursive: true, mode: "700" });
    await fs.writeFile(loc, JSON.stringify(data), { mode: "600" });
    return data;
  };

  try {
    const stat = await fs.stat(loc);
    if (stat.mtime.getTime() < Date.now() - options.duration) {
      await fs.rm(loc);
      return await loadCache();
    }

    const data = JSON.parse((await fs.readFile(loc)).toString("utf-8")) as T;
    if (hasExpired?.(data)) {
      await fs.rm(loc);
      return await loadCache();
    }
    return data;
  } catch (error: any) {
    if (error?.code !== "ENOENT")
      print2(
        `Could not load credentials "${name}" from cache: ${error.message ?? error}`
      );
    return await loadCache();
  }
};

const clearIdentityFile = async () => {
  try {
    const identityFilePath = getIdentityFilePath();
    // check to see if the file exists before trying to remove it
    await fs.access(identityFilePath);
    await fs.rm(identityFilePath);
  } catch {
    return;
  }
};

const clearIdentityCache = async () => {
  try {
    const identityCachePath = getIdentityCachePath();
    // check to see if the directory exists before trying to remove it
    await fs.access(identityCachePath);
    await fs.rm(identityCachePath, { recursive: true });
  } catch {
    return;
  }
};

const loadCredentialsWithAutoLogin = async (options?: {
  noRefresh?: boolean;
}): Promise<Identity> => {
  try {
    const buffer = await fs.readFile(getIdentityFilePath());
    const identity: Identity = JSON.parse(buffer.toString());
    if (
      !options?.noRefresh &&
      identity.credential.expires_at < Date.now() * 1e-3
    ) {
      await login({ org: identity.org.slug }, { skipAuthenticate: true });
      print2("\u200B"); // Force a new line
      return loadCredentialsWithAutoLogin({ noRefresh: true });
    }
    return identity;
  } catch (error: any) {
    if (error?.code === "ENOENT") {
      throw "Please run `p0 login <organization>` to use the P0 CLI.";
    }
    throw error;
  }
};

export const writeIdentity = async (
  org: OrgData,
  credential: TokenResponse
) => {
  await clearIdentityCache();

  const identityFilePath = getIdentityFilePath();

  const expires_at = Date.now() * 1e-3 + credential.expires_in - 1; // Add 1 second safety margin
  print2(`Saving authorization to ${identityFilePath}.`);
  const dir = path.dirname(identityFilePath);
  await fs.mkdir(dir, { recursive: true });
  await fs.writeFile(
    identityFilePath,
    JSON.stringify(
      {
        credential: { ...credential, expires_at },
        org,
      },
      null,
      2
    ),
    {
      mode: "600",
    }
  );
};

export const deleteIdentity = async () => {
  await clearIdentityCache();
  await clearIdentityFile();
};

export const authenticate = async (options?: {
  noRefresh?: boolean;
}): Promise<Authn> => {
  const identity = await loadCredentialsWithAutoLogin(options);
  const userCredential = await authenticateToFirebase(identity);

  return { userCredential, identity };
};
