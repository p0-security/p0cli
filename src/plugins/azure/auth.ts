/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { print2 } from "../../drivers/stdio";
import { exec } from "../../util";

export const azLoginCommand = () => ({
  command: "az",
  args: ["login"],
});

export const azLogoutCommand = () => ({
  command: "az",
  args: ["logout"],
});

export const azAccountSetCommand = (subscriptionId: string) => ({
  command: "az",
  args: ["account", "set", "--subscription", subscriptionId],
});

export const azLogin = async (
  subscriptionId: string,
  options: { debug?: boolean } = {}
) => {
  const { debug } = options;

  if (debug) print2("Logging in to Azure...");

  // Logging out first ensures that any cached credentials are cleared.
  // https://github.com/Azure/azure-cli/issues/29161
  try {
    const { command: azLogoutExe, args: azLogoutArgs } = azLogoutCommand();
    const logoutResult = await exec(azLogoutExe, azLogoutArgs, { check: true });

    if (debug) {
      print2(logoutResult.stdout);
      print2(logoutResult.stderr);
    }
  } catch (error: any) {
    if (debug) {
      // ignore the error if the user is not logged in.
      print2(`Skipping logout: ${error.stderr}`);
    }
  }

  const { command: azLoginExe, args: azLoginArgs } = azLoginCommand();
  const loginResult = await exec(azLoginExe, azLoginArgs, { check: true });

  if (debug) {
    print2(loginResult.stdout);
    print2(loginResult.stderr);
    print2(`Setting active Azure subscription to ${subscriptionId}...`);
  }

  const { command: azAccountSetExe, args: azAccountSetArgs } =
    azAccountSetCommand(subscriptionId);
  const accountSetResult = await exec(azAccountSetExe, azAccountSetArgs, {
    check: true,
  });

  if (debug) {
    print2(accountSetResult.stdout);
    print2(accountSetResult.stderr);
  }
};
