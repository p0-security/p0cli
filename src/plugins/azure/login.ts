/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { OIDC_HEADERS } from "../../common/auth/oidc";
import { withRedirectServer } from "../../common/auth/server";
import { urlEncode, validateResponse } from "../../common/fetch";
import { print2 } from "../../drivers/stdio";
import { AuthorizeRequest, TokenResponse } from "../../types/oidc";
import { OrgData } from "../../types/org";
import open from "open";
import pkceChallenge from "pkce-challenge";

const AZURE_SCOPE = "openid profile email offline_access";
const AZURE_REDIRECT_PORT = 52701;
const AZURE_REDIRECT_URL = `http://localhost:${AZURE_REDIRECT_PORT}`;
const PKCE_LENGTH = 128;

type CodeExchange = {
  code: string;
  state: string;
};

const requestAuth = async (org: OrgData) => {
  if (!org.providerDomain) {
    throw "Azure login requires a configured provider domain.";
  }

  const pkce = await pkceChallenge(PKCE_LENGTH);
  const baseUrl = `https://login.microsoftonline.com/${org.providerDomain}/oauth2/v2.0/authorize`;

  const authBody: AuthorizeRequest = {
    client_id: org.clientId,
    code_challenge: pkce.code_challenge,
    code_challenge_method: "S256",
    redirect_uri: AZURE_REDIRECT_URL,
    response_type: "code",
    scope: AZURE_SCOPE,
    state: "azure_login",
  };

  const url = `${baseUrl}?${urlEncode(authBody)}`;

  print2(`Your browser has been opened to visit:

    ${url}`);

  open(url).catch(() => {
    print2(`Please visit the following URL to continue login:

    ${url}`);
  });

  return pkce;
};

const requestToken = async (
  org: OrgData,
  code: string,
  pkce: { code_challenge: string; code_verifier: string }
) => {
  if (!org.providerDomain) {
    throw "Azure login requires a configured provider domain.";
  }

  const tokenUrl = `https://login.microsoftonline.com/${org.providerDomain}/oauth2/v2.0/token`;

  const body = {
    client_id: org.clientId,
    code,
    code_verifier: pkce.code_verifier,
    grant_type: "authorization_code",
    redirect_uri: AZURE_REDIRECT_URL,
  };

  const response = await fetch(tokenUrl, {
    method: "POST",
    headers: {
      ...OIDC_HEADERS,
      Accept: "application/json",
      Origin: AZURE_REDIRECT_URL,
    },
    body: urlEncode(body),
  });

  const valid = await validateResponse(response);
  return (await valid.json()) as TokenResponse;
};

export const azureLogin = async (org: OrgData): Promise<TokenResponse> => {
  return await withRedirectServer<any, CodeExchange, TokenResponse>(
    async () => await requestAuth(org),
    async (pkce, token) => await requestToken(org, token.code, pkce),
    { port: AZURE_REDIRECT_PORT }
  );
};
