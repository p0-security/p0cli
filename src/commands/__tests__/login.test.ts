import { pluginLoginMap } from "../../plugins/login";
import { mockGetDoc } from "../../testing/firestore";
import { login } from "../login";
import { signInWithCredential } from "firebase/auth";
import { readFile, writeFile } from "fs/promises";

jest.spyOn(Date, "now").mockReturnValue(1.6e12);
jest.mock("fs/promises");
jest.mock("../../drivers/auth", () => ({
  ...jest.requireActual("../../drivers/auth"),
  IDENTITY_FILE_PATH: "/path/to/home/.p0",
}));
jest.mock("../../drivers/stdio");
jest.mock("../../plugins/login");

const mockReadFile = readFile as jest.Mock;
const mockWriteFile = writeFile as jest.Mock;

describe("login", () => {
  it("prints a friendly error if the org is not found", async () => {
    mockGetDoc(undefined);
    await expect(login({ org: "test-org" })).rejects.toMatchInlineSnapshot(
      `"Could not find organization"`
    );
  });
  it("should print a friendly error if unsupported login", async () => {
    mockGetDoc({
      slug: "test-org",
      tenantId: "test-tenant",
      ssoProvider: "microsoft",
    });
    await expect(login({ org: "test-org" })).rejects.toMatchInlineSnapshot(
      `"Unsupported login for your organization"`
    );
  });
  describe("organization exists", () => {
    let credentialData: string = "";
    mockReadFile.mockImplementation(async () =>
      Buffer.from(credentialData, "utf-8")
    );
    mockWriteFile.mockImplementation(async (_path, data) => {
      credentialData = data;
    });
    beforeEach(() => {
      credentialData = "";
      jest.clearAllMocks();
      mockGetDoc({
        slug: "test-org",
        tenantId: "test-tenant",
        ssoProvider: "google",
      });
    });
    it("should call the provider's login function", async () => {
      await login({ org: "test-org" });
      expect(pluginLoginMap.google).toHaveBeenCalled();
    });
    it("should write the user's identity to the file system", async () => {
      await login({ org: "test-org" });
      expect(mockWriteFile.mock.calls).toMatchSnapshot();
    });
    it("validates authentication", async () => {
      await login({ org: "test-org" });
      expect((signInWithCredential as jest.Mock).mock.calls).toMatchSnapshot();
    });
  });
});
