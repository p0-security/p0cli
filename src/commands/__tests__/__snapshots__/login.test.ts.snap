// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login organization exists should write the user's identity to the file system 1`] = `
[
  [
    "/path/to/home/.p0",
    "{
  "credential": {
    "access_token": "test-access-token",
    "id_token": "test-id-token",
    "token_type": "oidc",
    "scope": "oidc",
    "expires_in": 3600,
    "refresh_token": "test-refresh-token",
    "device_secret": "test-device-secret",
    "expires_at": 1600003599
  },
  "org": {
    "slug": "test-org",
    "tenantId": "test-tenant",
    "ssoProvider": "google"
  }
}",
    {
      "mode": "600",
    },
  ],
]
`;

exports[`login organization exists validates authentication 1`] = `
[
  [
    {
      "tenantId": "test-tenant",
    },
    "test-credential",
  ],
]
`;
